harness/tests/cli/test_auth.py

```diff
@@ -29,74 +34,502 @@
+        return f"Use({self.user!r}, {self.token!r})"
 
 
 @contextlib.contextmanager
-def set_container_env_vars() -> Iterator[None]:
+def setenv_optional(key: str, val: Optional[str]) -> Iterator:
+    old = os.environ.get(key)
+    if val is None:
+        os.environ.pop(key, None)
+    else:
+        os.environ[key] = val
     try:
-        os.environ["DET_USER"] = "alice"
-        os.environ["DET_USER_TOKEN"] = "alice.token"
         yield
     finally:
-        del os.environ["DET_USER"]
-        del os.environ["DET_USER_TOKEN"]
+        if old is None:
+            os.environ.pop(key, None)
+        else:
+            os.environ[key] = old
+
+
+class Login(ScenarioSet):
+    req_user = {"y": "user", "n": None}
+    req_pass = {"y": "req_pass", "n": None}
+    env_user = {"y": "user", "e": "extra", "n": None}
+    env_pass = {"y": "env_pass", "n": None}
+    env_token = {"y": "env_token", "n": None}
+    cache = {"y": "cache", "x": "expired", "n": None}
+
+
+# Some shortcut results to save on line length.
+CheckCache = Check("cache")
+CheckExpired = Check("expired")
+UseCache = Use("user", "cache")
+UseNew = Use("user", "new")
+UseEnv = Use("user", "env_token")
 
 
-@pytest.mark.parametrize("user", [None, "bob", "determined"])
-@pytest.mark.parametrize("has_token_store", [True, False])
-def test_auth_user_from_env(
-    requests_mock: requests_mock.Mocker, user: Optional[str], has_token_store: bool
+@pytest.mark.parametrize(
+    "scenario_set",
+    [
+        #     requested user
+        #      |   requested password
+        #      |    |   DET_USER env var
+        #      |    |    |    DET_PASS env var
+        #      |    |    |     |   DET_USER_TOKEN
+        #      |    |    |     |    |   token cache state
+        #      |    |    |     |    |    |   Results...
+        #      |    |    |     |    |    |    |
+        #
+        # Explicit user and password, no environment settings, or DET_USER is overridden by
+        # explicitly requested user.
+        Login("y", "y", "ne", "*", "*", "y", CheckCache, UseCache),
+        Login("y", "y", "ne", "*", "*", "n", DoLogin("user", "req_pass"), UseNew),
+        Login("y", "y", "ne", "*", "*", "x", CheckExpired, DoLogin("user", "req_pass"), UseNew),
+        # ---
+        # Explicit user, but password not provided.  Still no (relevant) environment settings.
+        Login("y", "n", "ne", "*", "*", "y", CheckCache, UseCache),
+        Login("y", "n", "ne", "*", "*", "n", DoLogin("user", "prompt_pass"), UseNew),
+        Login("y", "n", "ne", "*", "*", "x", CheckExpired, DoLogin("user", "prompt_pass"), UseNew),
+        # ---
+        # Explicit user and password, DET_USER set but DET_USER_TOKEN not set.  DET_PASS is ignored.
+        Login("y", "y", "y", "*", "n", "y", CheckCache, UseCache),
+        Login("y", "y", "y", "*", "n", "n", DoLogin("user", "req_pass"), UseNew),
+        Login("y", "y", "y", "*", "n", "x", CheckExpired, DoLogin("user", "req_pass"), UseNew),
+        # ---
```

Comment:
I think we can add the following cases here:
 ```
        # Explicit user and password, DET_USER, DET_PASS, and DET_TOKEN set. DET_USER matches explicit.
        Login("y", "y", "y", "y", "y", "x", CheckExpired, UseEnv),
        Login("y", "y", "y", "*", "y", "n", UseEnv),
```

Reply:
second one is technically a duplicate of `DET_USER_TOKEN can be used if user is explicit`, but I am ok including it anyway because it makes the "everything is set" path pretty clear, and that is valuable.